Implementing output of EPS (and also PS) vector formats.

sauliusg opened the pull request:
As of current CDK tip, requesting PS output from Depiction (toEpsStr()) would trigger exception, which is confusing since PS is listed among supported format.
This PR implements output in PostScript (PS) and Encapsulated PostScript (EPS) formats; EPS output contains %%BoundingBox: as required by EPS specification. Such files are handy when including into larger documents, e.g. into LaTeX (PS workflow) generated files.

egonw commented on app/depict/src/main/java/org/openscience/cdk/depict/FreeHepWrapper.java:
```diff
@@ -98,13 +116,55 @@ public void dispose() {
 
+    // Documents cited below:
+    // [PLDS92] PostScript Language Document Structuring Conventions Specification, Version 3.0, 25 September 1992
+    // [EGFF96] J.D.Murray & W. vanPyper, Encyclopedia of Graphics File Formats 2nd ed., O'Reilly & Assoc., 1996
```
These citations may be better in the class JavaDoc and I would like to suggest to use the @cdk.cite tag for them.

egonw commented on base/silent/src/main/java/org/openscience/cdk/silent/SilentChemObjectBuilder.java:
```diff
@@ -143,7 +143,7 @@ private SilentChemObjectBuilder() {
             factory.register(IAtomContainer.class, AtomContainer.class);
         } else {
-            System.err.println("[INFO] Using the new AtomContainer implementation.");
+            // System.err.println("[INFO] Using the new AtomContainer implementation.");
```
This seems unrelated... maybe remove this change from this patch.

egonw commented on app/depict/src/main/java/org/openscience/cdk/depict/Depiction.java:
```diff
@@ -146,14 +151,24 @@ public final String toSvgStr(String units) {
     }
 
+    /**
+     * Render the image to an EPS (ncapsulated PostScript) format
```
Missing 'e' in 'ncapsulated'?

egonw commented on app/depict/src/main/java/org/openscience/cdk/depict/FreeHepWrapper.java:
```diff
@@ -98,13 +116,55 @@ public void dispose() {
         if (fmt.equals(Depiction.SVG_FMT)) {
             result = result.replaceAll("\"([-+0-9.]+)px\"", "\"$1mm\"");
         }
+        if (fmt.equals(Depiction.EPS_FMT)) {
```
The performance difference is probably minimal, but maybe use an 'else-if' here?

egonw commented on app/depict/src/test/java/org/openscience/cdk/depict/DepictionTest.java:
```diff
@@ -0,0 +1,49 @@
+package org.openscience.cdk.depict;
```
Please add the missing copyright/license header here.

sauliusg replied:
Added as of commit abbd8a4.

sauliusg commented:
Fixed most of the issues.

Please have a look if the @cdk.cite tag and JavaDocs are OK.

Cheers,
Saulius

On 2018-03-09 11:10, Egon Willighagen wrote:
> *@egonw* requested changes on this pull request.
> 
> Generally looks clear and clean. I left a few suggestions.
> 
> ------------------------------------------------------------------------
> 
> In
> app/depict/src/main/java/org/openscience/cdk/depict/FreeHepWrapper.java
> <https://github.com/cdk/cdk/pull/433#discussion_r173396175>:
> 
>> @@ -98,13 +116,55 @@ public void dispose() {
>          g2.dispose();
>      }
>  
> +    // Documents cited below:
> +    // [PLDS92] PostScript Language Document Structuring Conventions Specification, Version 3.0, 25 September 1992
> +    // [EGFF96] J.D.Murray & W. vanPyper, Encyclopedia of Graphics File Formats 2nd ed., O'Reilly & Assoc., 1996
> 
> These citations may be better in the class JavaDoc and I would like to
> suggest to use the @cdk.cite tag for them.
> 
> ------------------------------------------------------------------------
> 
> In
> base/silent/src/main/java/org/openscience/cdk/silent/SilentChemObjectBuilder.java
> <https://github.com/cdk/cdk/pull/433#discussion_r173396747>:
> 
>> @@ -143,7 +143,7 @@ private SilentChemObjectBuilder() {
>          if (CDK_LEGACY_AC) {
>              factory.register(IAtomContainer.class, AtomContainer.class);
>          } else {
> -            System.err.println("[INFO] Using the new AtomContainer implementation.");
> +            // System.err.println("[INFO] Using the new AtomContainer implementation.");
> 
> This seems unrelated... maybe remove this change from this patch.
> 
> ------------------------------------------------------------------------
> 
> In app/depict/src/main/java/org/openscience/cdk/depict/Depiction.java
> <https://github.com/cdk/cdk/pull/433#discussion_r173396896>:
> 
>>          return toVecStr(PS_FMT, UNITS_MM);
>      }
>  
> +    /**
> +     * Render the image to an EPS (ncapsulated PostScript) format
> 
> Missing 'e' in 'ncapsulated'?
> 
> ------------------------------------------------------------------------
> 
> In
> app/depict/src/main/java/org/openscience/cdk/depict/FreeHepWrapper.java
> <https://github.com/cdk/cdk/pull/433#discussion_r173397007>:
> 
>>      @Override
>      public String toString() {
>          String result = new String(bout.toByteArray(), StandardCharsets.UTF_8);
>          // we want SVG in mm not pixels!
>          if (fmt.equals(Depiction.SVG_FMT)) {
>              result = result.replaceAll("\"([-+0-9.]+)px\"", "\"$1mm\"");
>          }
> +        if (fmt.equals(Depiction.EPS_FMT)) {
> 
> The performance difference is probably minimal, but maybe use an
> 'else-if' here?
> 
> ------------------------------------------------------------------------
> 
> In
> app/depict/src/test/java/org/openscience/cdk/depict/DepictionTest.java
> <https://github.com/cdk/cdk/pull/433#discussion_r173397191>:
> 
>> @@ -0,0 +1,49 @@
> +package org.openscience.cdk.depict;
> 
> Please add the missing copyright/license header here.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cdk/cdk/pull/433#pullrequestreview-102580724>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKLJhGzp1UebtWln1H5wSGpcW-SaLBekks5tckcRgaJpZM4Sj8je>.
> 


-- 
Dr. Saulius Gražulis
Vilnius University Institute of Biotechnology, Saulėtekio al. 7
LT-10257 Vilnius, Lietuva (Lithuania)
fax: (+370-5)-2234367 / phone (office): (+370-5)-2234353
mobile: (+370-684)-49802, (+370-614)-36366

johnmay commented:
Need to look more carefully but generally looks okay...

> Such files are handy when including into larger documents, e.g. into LaTeX (PS workflow) generated files.

I just tend include the PDF structure output, is that bad/different?

johnmay commented:
> // can't scale page size correctly in FreeHEP atm

(my old comment)

Ah right I remember now why I took it out... had it kind of working but not everything so I just removed the code but forgot to remove the implementation. Thanks for the patch resolving it.

sauliusg commented:
>> Such files are handy when including into larger documents, e.g. into LaTeX (PS workflow) generated files.

> I just tend include the PDF structure output, is that bad/different?

PDF works with 'pdflatex' (PDF workflow) but must be converted for plain 'latex'. On the other hand, 'latex' works excellently with EPS, but does not handle PDF without conversion. Thus, being able to write out EPS can be handy when working with older software.

sauliusg commented:
>  // can't scale page size correctly in FreeHEP atm
> (my old comment)

> Ah right I remember now why I took it out... had it kind of working but not everything so I just removed the code but forgot to remove the implementation. Thanks for the patch resolving it.

You are right, setting page size seems to be broken in HEP. Also, %%BoundingBox is not handled by HEP library – we'll need to push a patch upstream, huh?

But for now, it seems possible to post-process an EPS string to have %%BoundingBox including; we get a valid EPS then. For EPS, page size is not relevant (the picture will get rescaled anyway). For PS, now a standard page size is always output, since HEP library seems to ignore size settings in Properties; but still, getting a standard-size PostScript is nicer than getting exception, IMHO :)

johnmay commented on app/depict/src/main/java/org/openscience/cdk/depict/FreeHepWrapper.java:
```diff
@@ -128,7 +128,17 @@ public String toString() {
+            // We should determine new-line separator (nl) not from OS type, but from the line-endings
+            // in the actual EPS outpu; there is nothing that would prevent us from generating Unix-style
+            // file in Windows :)
+            if( result.contains("\n\r")) {
```
'+1' was about to say when reading the previous commit you might need this. Good to see it's here but should'n't it be *\r\n* not *\n\r*. As a general rule of thumb is good to read both but only write one (e.g. our Molfile writer should only ever use \n - whether it does actually do that of course is another mater).

sauliusg replied:
You are write, will fix right now...

johnmay commented on app/depict/src/test/java/org/openscience/cdk/depict/DepictionTest.java:
```diff
@@ -1,3 +1,26 @@
+/*
+ * Copyright (c) 2015 John May <jwmay@users.sf.net>
```
Your name here

johnmay commented on base/silent/src/main/java/org/openscience/cdk/silent/SilentChemObjectBuilder.java:
```diff
@@ -143,7 +143,7 @@ private SilentChemObjectBuilder() {
         if (CDK_LEGACY_AC) {
             factory.register(IAtomContainer.class, AtomContainer.class);
         } else {
-            // System.err.println("[INFO] Using the new AtomContainer implementation.");
```
I'll fix that... meant to switch it round. Changed the flag recently should now be... "Using Legacy AtomContainer"

johnmay commented:
Two minor things left
 - please update the name on the copy right.
 - The **@cdk.cite** is incorrect

It's actually quite awkward and I actually prefer the simpler HREF as you've done... but semantically I guess it's better.

Step 1) - Add your citations to this BIBX file in a different project (cdk-build-util): ``https://github.com/cdk/cdk-build-util/blob/master/src/main/resources/net/sf/cdk/tools/doclets/cheminf.bibx``. Also we manually keep this in sync with the ``cheminf.bibx`` in the main project.
Step 2) - Now use the key you gave it in the BIBX file in an inline tag.

[Example](https://github.com/cdk/cdk/blob/master/base/isomorphism/src/main/java/org/openscience/cdk/isomorphism/VentoFoggia.java#L37-L38)

```java
/**
 * A structure pattern which utilises the Vento-Foggia (VF) algorithm {@cdk.cite
 * Cordella04}.
 */
```

This then gets inserted in a standard form when we build the doc

![image](https://user-images.githubusercontent.com/983232/37241465-7ec398de-2451-11e8-91f0-42db9c13cfd7.png)

johnmay commented:
I think there are already enough custom tags :-).

sauliusg commented:
> Step 1) - Add your citations to this BIBX file in a different project (cdk-build-util): https://github.com/cdk/cdk-build-util/blob/master/src/main/resources/net/sf/cdk/tools/doclets/cheminf.bibx. 

Added bibliographies and created PR to cdk-build-util – is this what you mean?

sauliusg commented:
> Step 1) - Add your citations to this BIBX file in a different project (cdk-build-util): https://github.com/cdk/cdk-build-util/blob/master/src/main/resources/net/sf/cdk/tools/doclets/cheminf.bibx.

BTW, the original cheminf.bibx fails to validate against bibtexml-20060727/schema/bibtexml-handwritten.xsd from http://bibtexml.sourceforge.net/ (in https://sourceforge.net/projects/bibtexml/files/BibTeXML/bibtexml-20060727.tar.bz2/download)...

sauliusg commented:
> Step 1) - Add your citations to this BIBX file in a different project (cdk-build-util): https://github.com/cdk/cdk-build-util/blob/master/src/main/resources/net/sf/cdk/tools/doclets/cheminf.bibx. Also we manually keep this in sync with the cheminf.bibx in the main project.
Step 2) - Now use the key you gave it in the BIBX file in an inline tag.

I've updated the references, but still not sure if this is how you do them in CDK.

IFAIU, the @cdk.cite tag works only in JavaDoc comments. You requested them at the beginning of the class. So I added them there.

However, I feel appropriate to cite the references right next to the code that implements them. So I left the original comments in-line, with the custom reference designators.

Or should I use {@cdk.cite} in the body of the method as well? Will they work in line comments ('//...')? Or should I change these comments to block ('/** ... */')? Will JavaDoc correctly find them inside the body of the method, or should I move them to the beginning of the method? Or to the beginning of the class?

Your help is greatly appreciated here.

Regards,
Saulius

egonw commented:
I'm happy with having the articles additionally in comments in the body... now they are visible to developers and users :)

johnmay commented:
Looks great thanks.

johnmay commented:
Was there many errors in the BIBX validator? As I said I don't really like the clunky-ness of it but we should try to make it valid.

sauliusg commented:
On 2018-03-11 15:19, John Mayfield wrote:

> Was there many errors in the BIBX validator?

In the current https://github.com/cdk/cdk-build-util master tip
(commit cbbd23f), there are 320 validation messages against
bibtexml-handwritten.xsd (from bibtexml-20060727.tar.bz2
MD5:398baf45c4ef7e87c2e288b058a3b8d3, obtained via
https://sourceforge.net/projects/bibtexml/files/BibTeXML/bibtexml-20060727.tar.bz2/download).

I've chosen 'bibtexml-handwritten.xsd' since it is the most complete;
another scheme, 'bibtexml-generated.xsd', is missing important elements
like 'url', 'size', 'abstract' and some others.

After suggested corrections (switching to 'bibtex:id', fixing encoding,
changing duplicate 'id' values, changing '<bibtex:doi>' to a supported
element, commenting out tree invalid element instances), there are only
8 validation messages left, all of them indicate missing <pages>,
<chapter> or <number> information.

Validation logs for both files are attached.

> As I said I don't really like the clunky-ness of it but we should
> try to make it valid.

An immediate way to reduce clutter is to specify
'xmlns="http://bibtexml.sf.net/"' in addition to 'xmlns:bibtex=...', and
remove all 'bibtex:' qualifiers from element names (they will still be
mandatory for 'bibtex:id' attributes, unless we change the schema).

Sincerely,
Saulius

-- 
Dr. Saulius Gražulis
Vilnius University Institute of Biotechnology, Saulėtekio al. 7
LT-10257 Vilnius, Lietuva (Lithuania)
fax: (+370-5)-2234367 / phone (office): (+370-5)-2234353
mobile: (+370-684)-49802, (+370-614)-36366
